test/extended/util/metrics.go

```diff
@@ -0,0 +1,327 @@
+			info.failed = true
+			podInfos = append(podInfos, info)
+			continue
+		}
+
+		procStartTime, ok := metrics["process_start_time_seconds"]
+		if !ok || len(procStartTime.Metric) == 0 {
+			info.result = fmt.Sprintf("error: failed to find metric: %q", "process_start_time_seconds")
+			info.failed = true
+			podInfos = append(podInfos, info)
+			continue
+		}
+		procStartTimeSeconds := procStartTime.Metric[0].GetGauge().GetValue()
+		totalProcTimeSeconds := time.Now().Unix() - int64(procStartTimeSeconds)
+
+		totalRequestCount := float64(0)
+		for _, metric := range metricGroup.Metric {
+			totalRequestCount += metric.Counter.GetValue()
+		}
+
+		comment := "within QPS bounds"
+		qps := totalRequestCount / float64(totalProcTimeSeconds)
+		if qps > maxQPSAllowed {
+			comment = "exceeds QPS bounds"
+		}
+		info.status = fmt.Sprintf("%s (%s)", info.status, comment)
+		info.qps = qps
+		info.result = fmt.Sprintf("%v requests over a span of %v seconds", totalRequestCount, totalProcTimeSeconds)
+		podInfos = append(podInfos, info)
+	}
+
+	return podInfos, nil
+}
+
+func getPodMetrics(adminConfig *restclient.Config, pod *v1.Pod, podUrlGetter *PortForwardURLGetter) (map[string]*dto.MetricFamily, error) {
+	result, err := podUrlGetter.Get("/metrics", pod, adminConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return parseRawMetrics(result)
+}
+
+func parseRawMetrics(rawMetrics string) (map[string]*dto.MetricFamily, error) {
+	p := expfmt.TextParser{}
+	return p.TextToMetricFamilies(bytes.NewBufferString(rawMetrics))
+}
+
+type defaultPortForwarder struct {
+	restConfig *rest.Config
+
+	StopChannel  chan struct{}
+	ReadyChannel chan struct{}
+}
+
+func NewDefaultPortForwarder(adminConfig *rest.Config) *defaultPortForwarder {
+	return &defaultPortForwarder{
+		restConfig:   adminConfig,
+		StopChannel:  make(chan struct{}, 1),
+		ReadyChannel: make(chan struct{}, 1),
+	}
+}
+
+func (f *defaultPortForwarder) ForwardPortsAndExecute(pod *v1.Pod, ports []string, toExecute func()) error {
+	if len(ports) < 1 {
+		return fmt.Errorf("at least 1 PORT is required for port-forward")
+	}
+
+	restClient, err := rest.RESTClientFor(setRESTConfigDefaults(*f.restConfig))
+	if err != nil {
+		return err
+	}
+
+	if pod.Status.Phase != v1.PodRunning {
+		return fmt.Errorf("unable to forward port because pod is not running. Current status=%v", pod.Status.Phase)
+	}
+
+	stdout := bytes.NewBuffer(nil)
+	req := restClient.Post().
+		Resource("pods").
+		Namespace(pod.Namespace).
+		Name(pod.Name).
+		SubResource("portforward")
+
+	transport, upgrader, err := spdy.RoundTripperFor(f.restConfig)
+	if err != nil {
+		return err
+	}
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, "POST", req.URL())
+	fw, err := portforward.New(dialer, ports, f.StopChannel, f.ReadyChannel, stdout, stdout)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		if f.StopChannel != nil {
```

Comment:
note to myself: think this channel handling trough

Reply:
:) inspiration for port-forward logic came from https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/portforward/portforward.go#L329